### PR TITLE
winx: do not link directly to ntdll.dll

### LINKED ISSUE #1738: Why does wasmtime runtime import setjmp from ntdll.dll?
I was looking into developing the Windows backend of Cranelift/wasmtime on Linux using cross-compilation and wine, and it actually works quite well, with one exception: I have to comment out the setjmp helper, because if I don't, this happens:

```
$ cargo run --target x86_64-pc-windows-gnu hello.wasm
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/x86_64-pc-windows-gnu/debug/wasmtime.exe hello.wasm`
wine: Call from 0x7bc5d60c to unimplemented function ntdll.dll._setjmp, aborting
wine: Unimplemented function ntdll.dll._setjmp called at address 0x7bc5d60c (thread 003e), starting debugger...
```

This clearly looks like a bug in wine of some sort, but given that setjmp is a pervasive function and it's neither implemented in wine nor can I find any similar bug reports, what wasmtime does here is at least *unusual*. What gives?

### PR DESCRIPTION
**Pull request by whitequark**
The documentation for `RtlNtStatusToDosError` explicitly tells to call it via `GetProcAddress`. The documentation for `NtQueryInformationFile` does not, but similar considerations apply because there is normally no import library for `ntdll`.

The main reason to use `GetProcAddress` though is because MinGW does include an import library for `ntdll`, this import library contains a definition of `setjmp`, and because of the way rustc orders linker arguments, this definition of `setjmp` conflicts with and gets picked over the one in `msvcrt`. Using `setjmp` from `ntdll` is undesirable as it is an undocumented API and it is missing from wine, making it harder to develop wasmtime using a cross-compiler.

Fixes #1738.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! Could this be done with an `AtomicUsize`, though, to avoid pulling in an extra dependency?

**Comment by whitequark** [comment]
That'll be kinda ugly, but I can do it, sure. Maybe define our own macro for ntdll imports.

**Comment by whitequark** [comment]
@alexcrichton Done

**Comment by whitequark** [comment]
I have no idea why the macOS test fails.

**Comment by kubkon** [comment]
> I have no idea why the macOS test fails.

Hmm, that's very weird indeed. I'll check out your branch locally on my mac and see if it trips for me locally as well.

File: crates/wasi-common/winx/src/ntdll.rs [hunk 1]
```diff
@@ -49,17 +51,39 @@ impl Default for IO_STATUS_BLOCK {
...
+        pub(crate) unsafe fn $name($($arg: $argty),*) -> $retty {
+            static ADDRESS: AtomicUsize = AtomicUsize::new(0);
+            let address = match ADDRESS.load(Ordering::Relaxed) {
+                0 => {
+                    let ntdll = GetModuleHandleA("ntdll\0".as_ptr() as *const i8);
+                    let address = GetProcAddress(
+                        ntdll,
+                        concat!(stringify!($name), "\0").as_ptr() as *const i8,
+                    ) as usize;
+                    ADDRESS.store(address, Ordering::Relaxed);
```
Comment (alexcrichton) [inline crates/wasi-common/winx/src/ntdll.rs:65]: Could you add an assert here that `address` isn't null as well?

Comment (whitequark) [reply crates/wasi-common/winx/src/ntdll.rs:65]: Done

**Comment by kubkon** [comment]
> > I have no idea why the macOS test fails.
> 
> Hmm, that's very weird indeed. I'll check out your branch locally on my mac and see if it trips for me locally as well.

@whitequark Ok, the CI issue seems to be random Github CI issue so hopefully next time you push, it should work itself out fine.

**Pull request merged**